zipline/data/us_equity_minutes.py

```diff
@@ -120,27 +119,56 @@ def _write_internal(self, directory, iterator, sid_path_func=None):
 
             os.makedirs(path)
 
-            minutes = self.full_minutes_for_days(_writer_env,
-                                                 first_open, df.index[-1])
+            last_dt = df.index[-1]
+
+            if all_minutes is None:
+                all_minutes = \
+                    self.full_minutes_for_days(first_open, last_dt)
+                minutes = all_minutes
+            else:
+                if df.index[-1] in all_minutes:
+                    mask = all_minutes.slice_indexer(end=last_dt)
+                    minutes = all_minutes[mask]
+                else:
+                    # Need to extend all minutes from open after last value
+                    # in all_minutes to the last_dt.
+                    next_open, _ = self._writer_env.next_open_and_close(
+                        all_minutes[-1])
+                    to_append = self.full_minutes_for_days(next_open, last_dt)
+                    all_minutes = all_minutes.append(to_append)
+                    minutes = all_minutes
+
             minutes_count = len(minutes)
 
-            dt_col = np.zeros(minutes_count, dtype=np.uint32)
             open_col = np.zeros(minutes_count, dtype=np.uint32)
             high_col = np.zeros(minutes_count, dtype=np.uint32)
             low_col = np.zeros(minutes_count, dtype=np.uint32)
             close_col = np.zeros(minutes_count, dtype=np.uint32)
             vol_col = np.zeros(minutes_count, dtype=np.uint32)
 
-            for row in df.iterrows():
-                dt = row[0]
-                idx = minutes.searchsorted(dt)
-
-                dt_col[idx] = dt.value / 1e9
-                open_col[idx] = row[1].loc["open"]
-                high_col[idx] = row[1].loc["high"]
-                low_col[idx] = row[1].loc["low"]
-                close_col[idx] = row[1].loc["close"]
-                vol_col[idx] = row[1].loc["volume"]
+            opens = df.open.values.astype(np.uint32)
+            highs = df.high.values.astype(np.uint32)
+            lows = df.low.values.astype(np.uint32)
+            closes = df.close.values.astype(np.uint32)
+            volumes = df.volume.values.astype(np.uint32)
+
+            dt_ixs = np.searchsorted(minutes.values, df.index.values)
+
+            for i, dt_ix in enumerate(dt_ixs):
```

Comment:
I think this is just ` open_col[dt_ixs] = opens` (with appropriate substitutions for the others).